[BUG] When AQE=on, SMJ with a Map in SELECTed list fails with "key not found: numPartitions"

**Describe the bug**
A clear and concise description of what the bug is.
When AQE=on, SMJ with a Map in SELECTed list fails with "key not found: numPartitions".

Full stacktrace is:
```
Caused by: java.util.NoSuchElementException: key not found: numPartitions
  at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:101)
  at scala.collection.immutable.Map$EmptyMap$.apply(Map.scala:99)
  at org.apache.spark.sql.execution.adaptive.CustomShuffleReaderExec.sendDriverMetrics(CustomShuffleReaderExec.scala:122)
  at org.apache.spark.sql.execution.adaptive.CustomShuffleReaderExec.shuffleRDD$lzycompute(CustomShuffleReaderExec.scala:182)
  at org.apache.spark.sql.execution.adaptive.CustomShuffleReaderExec.shuffleRDD(CustomShuffleReaderExec.scala:181)
  at org.apache.spark.sql.execution.adaptive.CustomShuffleReaderExec.doExecuteColumnar(CustomShuffleReaderExec.scala:196)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:207)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
  at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:203)
  at org.apache.spark.sql.execution.InputAdapter.doExecuteColumnar(WholeStageCodegenExec.scala:519)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:207)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
  at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:203)
  at org.apache.spark.sql.execution.ColumnarToRowExec.inputRDDs(Columnar.scala:202)
  at org.apache.spark.sql.execution.SortExec.inputRDDs(SortExec.scala:132)
  at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:746)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:180)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
  at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:176)
  at org.apache.spark.sql.execution.InputAdapter.doExecute(WholeStageCodegenExec.scala:511)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:180)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
  at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:176)
  at org.apache.spark.sql.execution.joins.SortMergeJoinExec.inputRDDs(SortMergeJoinExec.scala:378)
  at org.apache.spark.sql.execution.ProjectExec.inputRDDs(basicPhysicalOperators.scala:50)
  at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:746)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:180)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
  at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:176)
  at com.nvidia.spark.rapids.GpuRowToColumnarExec.doExecuteColumnar(GpuRowToColumnarExec.scala:836)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:207)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
  at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:203)
  at com.nvidia.spark.rapids.GpuBaseLimitExec.doExecuteColumnar(limit.scala:62)
  at com.nvidia.spark.rapids.GpuBaseLimitExec.doExecuteColumnar$(limit.scala:57)
  at com.nvidia.spark.rapids.GpuLocalLimitExec.doExecuteColumnar(limit.scala:123)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:207)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
  at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:203)
  at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.inputBatchRDD$lzycompute(GpuShuffleExchangeExec.scala:137)
  at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.inputBatchRDD(GpuShuffleExchangeExec.scala:137)
  at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.shuffleDependencyColumnar$lzycompute(GpuShuffleExchangeExec.scala:147)
  at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.shuffleDependencyColumnar(GpuShuffleExchangeExec.scala:145)
  at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.$anonfun$doExecuteColumnar$1(GpuShuffleExchangeExec.scala:169)
  at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
  ... 99 more
```

**Steps/Code to reproduce bug**
Please provide a list of steps or a code sample to reproduce the issue.
Avoid posting private or sensitive data.

Minimum repro:
```
import org.apache.spark.sql.Row
import org.apache.spark.sql.types._

val data = Seq(
    Row("Adam",Map("hair"->"black","eye"->"black"), Row("address" , Map("state"->"CA"))),
    Row("Bob",Map("hair"->"red","eye"->"red"),Row("address", Map("state"->"GA"))),
    Row("Cathy",Map("hair"->"blue","eye"->"blue"),Row("address",Map("state"->"NC")))
)

val mapType  = DataTypes.createMapType(StringType,StringType)

val schema = new StructType().add("name",StringType).add("properties", mapType)

val mapTypeDF = spark.createDataFrame(spark.sparkContext.parallelize(data),schema)
mapTypeDF.write.format("parquet").mode("overwrite").save("/tmp/testparquet")
val df1 = spark.read.parquet("/tmp/testparquet")
val df2 = spark.read.parquet("/tmp/testparquet")
df1.createOrReplaceTempView("df1")
df2.createOrReplaceTempView("df2")
df1.printSchema


spark.conf.set("spark.sql.adaptive.enabled",true)
spark.conf.set("spark.sql.autoBroadcastJoinThreshold",1)
val joindf=spark.sql("select df1.properties from df1, df2 where df1.name=df2.name")
joindf.show
joindf.explain

```


**Expected behavior**
A clear and concise description of what you expected to happen.

It should fall back on CPU nicely instead of failing.

**Environment details (please complete the following information)**
 - Environment location: [Standalone, YARN, Kubernetes, Cloud(specify cloud provider)]
 - Spark configuration settings related to the issue

Spark 3.1.1 standalone cluster
RAPIDS plugin 21.08 snapshot

**Additional context**
Add any other context about the problem here.

--- a/docs/supported_ops.md
+++ b/docs/supported_ops.md
@@ -443,9 +443,9 @@ Accelerator supports are described below.
 <td>S</td>
 <td><b>NS</b></td>
 <td><b>NS</b></td>
-<td><em>PS* (missing nested BINARY, CALENDAR, MAP, UDT)</em></td>
-<td><b>NS</b></td>
-<td><em>PS* (missing nested BINARY, CALENDAR, MAP, UDT)</em></td>
+<td><em>PS* (missing nested BINARY, CALENDAR, UDT)</em></td>
+<td><em>PS* (missing nested BINARY, CALENDAR, UDT)</em></td>
+<td><em>PS* (missing nested BINARY, CALENDAR, UDT)</em></td>
 <td><b>NS</b></td>
 </tr>
 <tr>

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
@@ -3048,7 +3048,7 @@ object GpuOverrides {
     exec[CustomShuffleReaderExec](
       "A wrapper of shuffle query stage",
       ExecChecks((TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL + TypeSig.ARRAY +
-        TypeSig.STRUCT).nested(), TypeSig.all),
+        TypeSig.STRUCT + TypeSig.MAP).nested(), TypeSig.all),
       (exec, conf, p, r) =>
       new SparkPlanMeta[CustomShuffleReaderExec](exec, conf, p, r) {
         override def tagPlanForGpu(): Unit = {

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala
@@ -113,7 +113,7 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
       val plan = GpuTransitionOverrides.getNonQueryStagePlan(s)
       if (plan.supportsColumnar && plan.isInstanceOf[GpuExec]) {
         parent match {
-          case Some(_: GpuCustomShuffleReaderExec) =>
+          case Some(_: GpuCustomShuffleReaderExec | _: CustomShuffleReaderExec) =>
             // We can't insert a coalesce batches operator between a custom shuffle reader
             // and a shuffle query stage, so we instead insert it around the custom shuffle
             // reader later on, in the next top-level case clause.

--- a/tests/src/test/scala/com/nvidia/spark/rapids/AdaptiveQueryExecSuite.scala
+++ b/tests/src/test/scala/com/nvidia/spark/rapids/AdaptiveQueryExecSuite.scala
@@ -32,7 +32,7 @@ import org.apache.spark.sql.execution.joins.SortMergeJoinExec
 import org.apache.spark.sql.functions.{col, when}
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.rapids.execution.GpuCustomShuffleReaderExec
-import org.apache.spark.sql.types.{ArrayType, DecimalType, IntegerType, StructField, StructType}
+import org.apache.spark.sql.types.{ArrayType, DataTypes, DecimalType, IntegerType, StringType, StructField, StructType}
 
 object AdaptiveQueryExecSuite {
   val TEST_FILES_ROOT: File = TestUtils.getTempDir(this.getClass.getSimpleName)
@@ -493,6 +493,35 @@ class AdaptiveQueryExecSuite
       }
       // Verify local readers length
       assert(localReaders2.length == 2)
+    }, conf)
+  }
+
+  test("Avoid splitting CustomShuffleReader and ShuffleQueryStageExec pairs") {
+    logError("Avoid splitting CustomShuffleReader and ShuffleQueryStageExec pairs")
+
+    val conf = new SparkConf()
+        .set(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key, "true")
+        .set(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key, "1")
+        .set(RapidsConf.ENABLE_REPLACE_SORTMERGEJOIN.key, "false")
+        .set("spark.rapids.sql.exec.SortExec", "false")
+        .set("spark.rapids.sql.exec.SortMergeJoinExec", "false")
+        .set(RapidsConf.TEST_ALLOWED_NONGPU.key,
+          "DataWritingCommandExec,ShuffleExchangeExec,HashPartitioning,SortExec,SortMergeJoinExec")
+
+    withGpuSparkSession(spark => {
+      val data = Seq(
+        Row("Adam",Map("hair"->"black","eye"->"black"), Row("address" , Map("state"->"CA"))),
+        Row("Bob",Map("hair"->"red","eye"->"red"),Row("address", Map("state"->"GA"))),
+        Row("Cathy",Map("hair"->"blue","eye"->"blue"),Row("address",Map("state"->"NC")))
+      )
+      val mapType  = DataTypes.createMapType(StringType,StringType)
+      val schema = new StructType().add("name",StringType).add("properties", mapType)
+      val df = spark.createDataFrame(spark.sparkContext.parallelize(data),schema)
+      df.createOrReplaceTempView("df1")
+      df.createOrReplaceTempView("df2")
+
+      val (plan, adaptivePlan) = runAdaptiveAndVerifyResult(spark,
+        "SELECT df1.properties from df1, df2 where df1.name=df2.name")
     }, conf)
   }
 

